### PR TITLE
Reduce string allocations for logevents with single target destination

### PR DESCRIPTION
**Pull request by snakefoot**
Will make some performance tests later to see how it performs.

Should also reduce pressure on the internal StringBuilderPool, as it will just write directly to the final destination-StringBuilder.

### CODE REVIEW

File: src/NLog/Internal/LogMessageTemplateFormatter.cs [hunk 1]
```diff
@@ -35,70 +35,92 @@ namespace NLog.Internal
...
         {
-            //if message is empty, there no parameters
-            //null check cheapest, so in-front
+            if (HasParameters(logEvent))
+            {
+                if (_singleTargetOnly)
+                {
+                    // Make quick check for valid message template parameter names
+                    TemplateEnumerator holeEnumerator = new TemplateEnumerator(logEvent.Message);
+                    if (holeEnumerator.MoveNext() && holeEnumerator.Current.Literal.Skip != 0 && holeEnumerator.Current.Hole.Index != -1)
```
Comment (304NotModified) [inline src/NLog/Internal/LogMessageTemplateFormatter.cs]: Micro-optimization, but I assume the int compare is a lot faster than .MoveNext

Comment (snakefoot) [reply src/NLog/Internal/LogMessageTemplateFormatter.cs]: But I need to MoveNext() to have a Current. Or am I missing something?

**Comment by 304NotModified** [comment]
I am curious about the results :) I think could be a nice improvement

File: src/NLog/Internal/LogMessageTemplateFormatter.cs [hunk 2]
```diff
@@ -35,70 +35,92 @@ namespace NLog.Internal
...
         public LogMessageFormatter MessageFormatter { get; }
-
+ 
         public bool HasProperties(LogEventInfo logEvent)
         {
-            //if message is empty, there no parameters
-            //null check cheapest, so in-front
+            if (HasParameters(logEvent))
+            {
+                if (_singleTargetOnly)
```
Comment (304NotModified) [inline src/NLog/Internal/LogMessageTemplateFormatter.cs]: micro: also _singleTargetOnly is cheaper to check than HasParameters?

File: src/NLog/Internal/LogMessageTemplateFormatter.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Internal/LogMessageTemplateFormatter.cs]: nope, I missed it - oops

File: src/NLog/Internal/LogMessageTemplateFormatter.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Internal/LogMessageTemplateFormatter.cs]: Will refactor the code, so it is more obvious what I'm trying to do.

**Comment by snakefoot** [comment]
Have executed some performance tests:

- **Single ZeroArgTest** - Single logger-thread that skips optimization
  - No huge impact, from checking if optimization can be used.
- **Single DeferTest** - Single logger-thread, that defers string-formatting to target-thread
  - Optimizations now cancels the overhead introduced with NLog 4.5, caused by always scanning the message template for structured parameters.
- **Single PreCalcTest** - Single logger-thread must execute string-formatting before giving to target-thread
  - Optimization reduces the cpu- and memory-cost for the logger-thread
- **Multi PreCalcTest** - Multiple logger-threads must execute string-formatting before giving to target-thread
  - Optimization reduces the memory-cost for the logger-threads
- **Multi DeferTest** - Single logger-thread thread that can defer formatting to multiple target-threads (No capture needed)
  - Optimization reduces the cpu- and memory-cost for the target-threads

| Single ZeroArgTest      | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| PerfLogger       | 10.000.000 |   64 |    0 |       1 |       1 |

| Single ZeroArgTest  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    7.885 | 1.268.209 |   0 |  14 |  29 |   10.125 |    1.489,9 |
| Master    |    6.627 | 1.508.855 |   0 |  13 |  27 |    8.984 |    1.451,7 |
| This PR   |     6.634 | 1.507.185 |   0 |  13 |  28 |    9.015 |    1.451,7 |

| Single DeferTest | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| PerfLogger        | 10.000.000 |   64 |    1 |       1 |       1 |

| Single DeferTest | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    12.894 |   775.498 |   0 |  23 |  47 |   16.421 |    2.862,9 |
| Master    |    13.857 |   721.618 |   0 |  25 |  52 |   17.250 |    2.824,7 |
| This PR    |    12.947 |   772.341 |   0 |  11 |  26 |   16.328 |    1.450,7 |

| Single PreCalcTest | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| Logger           | 10.000.000 |   64 |    1 |       1 |       1 |

| Single PreCalcTest | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    24.966 |   400.535 |   0 |  37 |  75 |   44.437 |    7.353,4 |
| Master    |    22.231 |   449.820 |   0 |  58 | 119 |   42.343 |    6.589,1 |
| This PR    |    21.007 |   476.011 |   1 |  45 |  91 |   41.250 |    5.216,4 |

| Multi PreCalcTest | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| Logger           | 10.000.000 |   64 |    1 |       4 |       1 |

| Multi PreCalcTest | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    41.582 |   240.484 |   0 |  34 |  69 |  106.265 |    7.372,9 |
| Master    |   34.931 |   286.273 |   0 |  31 |  63 |  102.703 |    6.606,2 |
| This PR    |    35.049 |   285.306 |   0 |  25 |  52 |  101.734 |    5.231,6 |

| Multi DeferTest | Messages   | Size | Args | Threads | Loggers |
|------------------|------------|------|------|---------|---------|
| MultiLogger      | 10.000.000 |   64 |    1 |       1 |       1 |

| Multi DeferTest | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog 4.4    |    8.042 | 1.243.468 |   4 |  20 |  41 |   28.546 |    3.610,3 |
| Master    |    8.083 | 1.237.125 |   2 |  14 |  29 |   29.625 |    2.840,6 |
| This PR    |    7.053 | 1.417.732 |   1 |   6 |  13 |   26.453 |    1.459,2 |

**Comment by 304NotModified** [comment]
nice work! so 5-12% and/or reduced GC :)

**Pull request merged**